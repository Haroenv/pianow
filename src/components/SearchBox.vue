<template lang="pug">
.search-box
  form(@submit.prevent="submit")
    div(role="search")
      label.search-label(for="search") Search
        p(id="error" role="alert")
      input#search.search-input(
        ref="input"
        type="text"
        :placeholder="placeholder"
        :value="value"
        @focus="inputFocus"
        @blur="inputBlur"
        @input="updateValue($event.target.value)"
        @keyup.enter="inputKeyEnter"
        @keydown.tab="inputKeyEnter" 
        @keydown.up="inputKeyUp" 
        @keydown.down="inputKeyDown"
        )
      input.button(type="submit" value="Search")
      button(type="button" @click="updateValue('')") reset
    .search-box__autocomplete(v-if="autocomplete" v-show="showAutocomplete" @click.stop="autocompleteFocus")
      slot
        search-results(v-on:hit:selected="updateValue($event.name)")
</template>

<script>
import algoliasearchHelper from "algoliasearch-helper";
import { setTimeout } from "timers";

export default {
  name: "SearchBox",
  inject: ["_client", "_helper", "_index"],
  props: {
    value: {
      type: String
    },
    autocomplete: {
      type: Boolean,
      default: false
    },
    placeholder: {
      type: String,
      default: ""
    },
    options: {
      type: Object
    }
  },
  data() {
    return {
      active: null,
      links: [],
      linkIndex: -1,
      hasInputFocus: false,
      hasAutoCompleteFocus: false,
      localHelper: null,
      hits: []
    };
  },
  created() {
    // blur when clicking outside
    document.body.addEventListener("click", this.autocompleteBlur);
  },
  destroyed() {
    // remove click outside event
    document.body.removeEventListener("click", this.autocompleteBlur);
  },
  computed: {
    showAutocomplete() {
      return this.hasFocus && this.value.length > 0;
    },
    hasFocus() {
      return this.hasInputFocus || this.hasAutoCompleteFocus;
    }
  },
  methods: {
    inputBlur() {
      // delay so hasAutoCompleteFocus has time to trigger
      setTimeout(() => (this.hasInputFocus = false), 200);
    },
    inputFocus() {
      this.autocompleteReset();
      this.hasInputFocus = true;
    },
    setActiveLink() {
      // remove existing active link
      if (this.active) {
        this.active.classList.remove("active");
        this.active = null;
      }
      // select new link
      if (this.linkIndex >= 0) {
        this.links[this.linkIndex].classList.add("active");
        this.active = this.$el.getElementsByClassName("active")[0];
      }
    },
    inputKeyUp() {
      if (this.linkIndex - 1 >= 0) {
        this.linkIndex--;
        this.setActiveLink();
      }
    },
    inputKeyDown() {
      if (this.linkIndex + 1 < this.links.length) {
        this.linkIndex++;
        this.setActiveLink();
      }
    },
    inputKeyEnter() {
      // trigger click on active element
      if (this.hasFocus && this.active) {
        this.active.click();
      }
    },
    autocompleteReset() {
      // reset link active
      this.linkIndex = -1;
      this.setActiveLink();
    },
    autocompleteBlur() {
      this.hasAutoCompleteFocus = false;
    },
    autocompleteFocus() {
      this.hasAutoCompleteFocus = true;
    },
    submit() {
      // trigger parent index search if autocomplete mode
      if (this.autocomplete) {
        this._helper.setQuery(this.value).search();
      }
    },
    updateValue(value) {
      this.$emit("input", value);
    }
  },
  watch: {
    // whenever input search value changes
    value(newValue, oldValue) {
      this.autocompleteBlur();
      this.autocompleteReset();

      // recompute links array
      this.links = this.$el.getElementsByClassName("search-results__link");

      // trigger search
      this.localHelper.setQuery(newValue).search();
    }
  },
  provide() {
    this.localHelper = this.autocomplete
      ? // create a separate helper for independant triggering
        algoliasearchHelper(this._client, this._index, this.options)
      : // No autocomplete, filter parent index directly
        this._helper;
    return {
      _helper: this.localHelper
    };
  }
};
</script>

<style scoped>
.search-box__autocomplete {
  position: absolute;
  background-color: grey;
  border: solid thin black;
}
</style>
