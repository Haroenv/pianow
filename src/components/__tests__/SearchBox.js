import Vue from "vue";
import SearchBox from "../SearchBox";

function createVm(
  helper,
  index,
  value = undefined,
  placeholder,
  searchLabel = undefined,
  searchButtonLabel = undefined,
  resetButtonLabel = undefined
) {
  return new Vue({
    components: {
      SearchBox
    },
    data() {
      return {
        value: value || "",
        index,
        placeholder,
        searchLabel,
        searchButtonLabel,
        resetButtonLabel
      };
    },
    provide: {
      _helper: helper,
      _index: "parentIndex"
    },
    template: `
    <search-box 
      v-model="value"
      :index="index"
      :placeholder="placeholder"
      :search-label="searchLabel"
      :search-button-label="searchButtonLabel"
      :reset-button-label="resetButtonLabel"
      />`
  }).$mount();
}

describe("SearchBox.vue", () => {
  const search = jest.fn();
  const helper = {
    setQuery: jest.fn(() => ({
      search: () => search()
    }))
  };

  beforeEach(() => {
    search.mockClear();
    helper.setQuery.mockClear();
  });

  it("should call parent helper setQuery and search when submitting the form", () => {
    const vm = createVm(helper, "testIndex", "test");
    const child = vm.$children[0];
    child.submit();
    expect(helper.setQuery).toHaveBeenCalledTimes(1);
    expect(helper.setQuery).toBeCalledWith("test");
  });

  it("should call parent helper setQuery and search on value changes", next => {
    const vm = createVm(helper, "testIndex");
    vm.value = "test";
    vm.$nextTick(() => {
      expect(helper.setQuery).toHaveBeenCalledTimes(1);
      expect(helper.setQuery).toBeCalledWith("test");
      next();
    });
  });

  it("should show correct content with default value", () => {
    const vm = createVm(helper, "testIndex");
    expect(vm.$el.querySelector(".search-box__input").placeholder).toEqual("");
    expect(vm.$el.querySelector(".search-box__label").textContent).toEqual(
      "Search"
    );
    expect(vm.$el.querySelector(".search-box__submit-button").value).toEqual(
      "Search"
    );
    expect(vm.$el.querySelector(".search-box__reset-button").value).toEqual(
      "Reset"
    );
  });

  it("should show correct content with custom value", () => {
    const vm = createVm(
      helper,
      "testIndex",
      "",
      "custom placeholder",
      "Custom search 1",
      "Custom search 2",
      "Custom reset"
    );
    expect(vm.$el.querySelector(".search-box__input").placeholder).toEqual(
      "custom placeholder"
    );
    expect(vm.$el.querySelector(".search-box__label").textContent).toEqual(
      "Custom search 1"
    );
    expect(vm.$el.querySelector(".search-box__submit-button").value).toEqual(
      "Custom search 2"
    );
    expect(vm.$el.querySelector(".search-box__reset-button").value).toEqual(
      "Custom reset"
    );
  });

  it("should match the snapshot", () => {
    const vm = createVm(helper, "testIndex");
    expect(vm.$el).toMatchSnapshot();
  });
});
