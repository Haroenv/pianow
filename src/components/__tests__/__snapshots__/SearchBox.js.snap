// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox.vue should match the snapshot 1`] = `
<div
  class="search-box"
  index="testIndex"
>
  <form>
    <div
      class="search-box__container"
      role="search"
    >
      <label
        class="search-box__label"
        for="search"
      >
        Search
      </label>
      <input
        class="search-box__input"
        id="search"
        placeholder=""
        type="text"
      />
      <input
        class="search-box__submit-button button"
        type="submit"
        value="Search"
      />
      <input
        class="search-box__reset-button button"
        type="reset"
        value="Reset"
      />
    </div>
  </form>
</div>
`;
